src/NumericalAlgorithms/RootFinding/GslMultiRoot.hpp

```diff
@@ -0,0 +1,364 @@
+// family of possible functions described in a class, but we instead pass the
+// pointer to the entire function object itself here. The
+// information of the function body is passed through with the Function template
+// parameter.
+template <size_t Dim, typename Function>
+int gsl_multirootfunctionfdf_wrapper_f(const gsl_vector* const x,
+                                       void* function_object,
+                                       gsl_vector* const f_of_x) noexcept {
+  const auto fn = static_cast<const Function*>(function_object);
+  gsl_vector_set_with_std_array(f_of_x,
+                                fn->operator()(gsl_to_std_array<Dim>(x)));
+  return GSL_SUCCESS;
+}
+
+template <size_t Dim, typename Function>
+int gsl_multirootfunctionfdf_wrapper_df(const gsl_vector* const x,
+                                        void* function_object,
+                                        gsl_matrix* const jacobian) noexcept {
+  const auto fn = static_cast<const Function*>(function_object);
+  gsl_matrix_set_with_std_array(jacobian,
+                                fn->jacobian(gsl_to_std_array<Dim>(x)));
+
+  return GSL_SUCCESS;
+}
+
+template <size_t Dim, typename Function>
+int gsl_multirootfunctionfdf_wrapper_fdf(const gsl_vector* const x,
+                                         void* function_object,
+                                         gsl_vector* const f_of_x,
+                                         gsl_matrix* const jacobian) noexcept {
+  const auto fn = static_cast<const Function*>(function_object);
+  const std::array<double, Dim> x_as_std_array = gsl_to_std_array<Dim>(x);
+  gsl_vector_set_with_std_array(f_of_x, fn->operator()(x_as_std_array));
+  gsl_matrix_set_with_std_array(jacobian, fn->jacobian(x_as_std_array));
+  return GSL_SUCCESS;
+}
+}  // end namespace detail
+
+/*!
+ *  \ingroup NumericalAlgorithmsGroup
+ *  \brief The exception thrown when gsl_multiroot fails to converge.
+ */
+class rootfinding_exception : public std::exception {
+  const char* what() const noexcept override { return "gsl_multiroot failed."; }
+};
+
+/*!
+ *  \ingroup NumericalAlgorithmsGroup
+ *  \brief The different options for the verbosity of gsl_multiroot.
+ */
+enum class Verbosity {
+  /// Do not print anything
+  Silent,
+  /// Print only "success" or "failed" on termination
+  Quiet,
+  /// Print final functions values on termination
+  Verbose,
+  /// Print function values on every iteration
+  Debug
+};
+
+/*!
+ *  \ingroup NumericalAlgorithmsGroup
+ *  \brief The different options for the rootfinding method of gsl_multiroot.
+ *
+ *  These methods require the function to have a jacobian provided: Hybrids,
+ *  Hybrid, and Newton. These methods do not require a provided jacobian:
+ *  Hybrids, Hybrid, DNewton.
+ */
+enum class Method {
+  /// Hybrid of Newton's method along with
+  /// following the gradient direction.
+  /// Note: Sometimes Hybrids works only
```

Comment:
How do I know when I've encountered one of these cases? Use `\note` for Doxygen notes. Maybe this should be a warning even?